### ISSUE TITLE
Bad diagnostics when using async fn's and attributes

### ISSUE THREAD

**Issue by matheus-consoli** [rust-lang/rust#102823]
[rustexplorer](https://www.rustexplorer.com/b#%2F*%0A%5Bdependencies%5D%0Atracing%20%3D%20%220.1.36%22%0A*%2F%0A%0A%23%5Bderive(Debug)%5D%0Astruct%20Foo%3B%0A%0Aimpl%20Foo%20%7B%0A%20%20%20%20%23%5Btracing%3A%3Ainstrument()%5D%0A%20%20%20%20async%20fn%20missing_implementation(%26self)%20-%3E%20char%20%7B%7D%0A%0A%20%20%20%20%23%5Btracing%3A%3Ainstrument()%5D%0A%20%20%20%20async%20fn%20missing_wrapper(%26self)%20-%3E%20Option%3Cchar%3E%20%7B%0A%20%20%20%20%20%20%20%20'a'%0A%20%20%20%20%7D%0A%0A%20%20%20%20%23%5Btracing%3A%3Ainstrument()%5D%0A%20%20%20%20fn%20missing_implementation_sync(%26self)%20-%3E%20char%20%7B%7D%0A%0A%20%20%20%20%23%5Btracing%3A%3Ainstrument()%5D%0A%20%20%20%20fn%20missing_wrapper_sync(%26self)%20-%3E%20Option%3Cchar%3E%20%7B%0A%20%20%20%20%20%20%20%20'a'%0A%20%20%20%20%7D%0A%7D%0A%0Afn%20main()%20%7B%7D) (rustexplorer is a playground alternative, I'm using it because play.rust-lang.org doesn't exposes `tracing` attributes)

This is probably related to some other issues, like #71968, but I decided to open it either way because the diagnostics differ from the async and sync version.

Simplifying my real case, given some struct:

```rust
#[derive(Debug)]
struct Foo;
```

I've tried implementing some instrumented associated async fn and hit check before finishing the code:

```rust
impl Foo {
    #[tracing::instrument()]
    async fn missing_implementation(&self) -> char {}

    #[tracing::instrument()]
    async fn missing_wrapper(&self) -> Option<char> {
        'a'
    }
}
```

which gave me weird diagnostics, especially the last one:

```rust
  --> src/main.rs:10:5
   |
10 |     #[tracing::instrument()]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ expected `char`, found `()`

error[E0308]: mismatched types
  --> src/main.rs:13:5
   |
13 |     #[tracing::instrument()]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `Option`, found `char`
   |
   = note: expected enum `Option<char>`
              found type `char`
help: try wrapping the expression in `Some`
   |
13 |     Some(#[tracing::instrument()])
   |     +++++                        +
```

but, to my surprise, testing with the sync version:

```rust
impl Foo {
    #[tracing::instrument()]
    fn missing_implementation(&self) -> char {}

    #[tracing::instrument()]
    fn missing_wrapper(&self) -> Option<char> {
        'a'
    }
}
```

outputs:

```rust
error[E0308]: mismatched types
  --> src/main.rs:19:51
   |
19 |     fn missing_implementation(&self) -> char {}
   |                                              ^^ expected `char`, found `()`

error[E0308]: mismatched types
  --> src/main.rs:23:9
   |
22 |     fn missing_wrapper(&self) -> Option<char> {
   |                                  ------------ expected `Option<char>` because of return type
23 |         'a'
   |         ^^^ expected enum `Option`, found `char`
   |
   = note: expected enum `Option<char>`
              found type `char`
help: try wrapping the expression in `Some`
   |
23 |         Some('a')
   |         +++++   +
```

I expect the async version outputs the same diagnostics as the sync version.

**Comment by compiler-errors** [comment]
@matheus-consoli this isn't really a rustc bug, but a bug in the proc macro expansion of tracing. I think I fixed it in tokio-rs/tracing#2270, can you try updating your tracing version and seeing if that fixes it? Thanks!

**Comment by matheus-consoli** [comment]
oh, sorry for opening it in the wrong repo!

it's fixed, thank you for your work
